stake-o-matic/src/main.rs

```diff
@@ -562,6 +562,64 @@ where
     }
 }
 
+fn classify_producers(
+    first_slot: Slot,
+    first_slot_in_epoch: Slot,
+    confirmed_blocks: HashSet<u64>,
+    leader_schedule: HashMap<String, Vec<usize>>,
+    quality_block_producer_percentage: usize,
+) -> Result<(HashSet<Pubkey>, HashSet<Pubkey>), Box<dyn error::Error>> {
+    let mut poor_block_producers = HashSet::new();
+    let mut quality_block_producers = HashSet::new();
+    let mut blocks_and_slots = HashMap::new();
+
+    let mut total_blocks = 0;
+    let mut total_slots = 0;
+    for (validator_identity, relative_slots) in leader_schedule {
+        let mut validator_blocks = 0;
+        let mut validator_slots = 0;
+        for relative_slot in relative_slots {
+            let slot = first_slot_in_epoch + relative_slot as Slot;
+            if slot >= first_slot {
+                total_slots += 1;
+                validator_slots += 1;
+                if confirmed_blocks.contains(&slot) {
+                    total_blocks += 1;
+                    validator_blocks += 1;
+                }
+            }
+        }
+        trace!(
+            "Validator {} produced {} blocks in {} slots",
+            validator_identity,
+            validator_blocks,
+            validator_slots
+        );
+        if validator_slots > 0 {
+            let validator_identity = Pubkey::from_str(&validator_identity)?;
+            let e = blocks_and_slots.entry(validator_identity).or_insert((0, 0));
+            e.0 += validator_blocks;
+            e.1 += validator_slots;
+        }
+    }
+    let cluster_average_rate = 100 - total_blocks * 100 / total_slots;
+    for (validator_identity, (blocks, slots)) in blocks_and_slots {
+        let skip_rate: usize = 100 - (blocks * 100 / slots);
+        if skip_rate.saturating_sub(quality_block_producer_percentage) >= cluster_average_rate {
+            poor_block_producers.insert(validator_identity);
+        } else {
+            quality_block_producers.insert(validator_identity);
+        }
+    }
+
+    info!("quality_block_producers: {}", quality_block_producers.len());
+    trace!("quality_block_producers: {:?}", quality_block_producers);
+    info!("poor_block_producers: {}", poor_block_producers.len());
+    trace!("poor_block_producers: {:?}", poor_block_producers);
+
+    Ok((quality_block_producers, poor_block_producers))
```

Comment:
Can we return `cluster_average_rate` as well and notify on it beyond some threshold?  This is for visibility more than anything else. :lobster:

Reply:
Like just doing a `info!` if it's above 50% or something?

Reply:
That might be enough, but I was thinking of returning it back to main and `notifier.send()`'ing it, so we see on Discord

Reply:
Around here would be good, https://github.com/solana-labs/solana/blob/24af9961d2cd9acbafca4869535b6c18346d0a0f/stake-o-matic/src/main.rs#L1517-L1523

Reply:
oh ok, I see.